python/lsst/ci/models.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import List, Set, Optional, Dict
+
+from lsst.ci import tsort
+
+DEFAULT_BRANCH_NAME = "master"
+
+
+class ProductIndex(dict):
+    """Container class for a product index with index-related operations,
+    such as topological sorting.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.toposorted = False
+        self.sorted_groups = []
+
+    def toposort(self) -> Dict[str, Product]:
+        """Topologically sort the product index and return it.
+
+        This mutates this object if it is not already sorted, and it returns
+        itself for convenience.
```

Comment:
Is it necessary to have this method modify self? It's a much cleaner API to return the sorted version and leave the original alone.

Another idea is to always have the data topologically sorted -- create the object that way and update the sort when any item is added. But if sorting is expensive and you plan to add a lot of items then that may be too expensive.